### ISSUE TITLE
Does 3D prediction fail if the `Z` dimension is too small (not a power of 2, less than 8)?

### ISSUE THREAD

**Issue by jdeschamps** [CAREamics/careamics#296]
## Question

We should test whether images with low number of slices (e.g. shape `(5, 128, 128)`) fail when predicting because of the `Z` dimensions not being power of 2 and less than 8?

If they do, would padding be a work-around?

**Comment by jdeschamps** [comment]
Also downsampling is applied similarly in `Z` as in `YX`, which might be suboptimal given that dimensions and resolutions are often different in `Z`.

**Comment by mese79** [comment]
Yes, It will fail if the `Z` dimension of a test stack is less than `8`.   
Especially for the BMZ export which calls the `predict` with the default config params ([link](https://github.com/CAREamics/careamics/blob/6a8a8ba016df5d866c3c739a2d4832f44ded60a8/src/careamics/careamist.py#L916)).   

I'm wondering if this is the model (algorithm) limitation, or just the CAREamics api ?

**Comment by jdeschamps** [comment]
Questions to solve:

- Where does the constraint come from (UNet?)
- We should guard against it by warning users before
- Can we fix it by reflecting boundaries?

**Comment by mese79** [comment]
The error on BMZ export for input of (1, 256, 256) is:
`RuntimeError: Given input size: (32x1x256x256). Calculated output size: (32x0x128x128). Output size is too small`
throwing from `torch.max_pool3d(input, kernel_size, stride, padding, dilation, ceil_mode)`.

It's working for the input of `(8, 256, 256)`.

**Comment by mese79** [comment]
I think I've found a solution (workaround) for this. For the `UNet` models, the first layer is a `MaxPool3D` which have no learnable parameter and basically does the `max` across the dimensions (using a kernel).  
```python
FCNModule(
  (model): UNet(
    (encoder): UnetEncoder(
      (pooling): MaxPool3d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
      (encoder_blocks): ModuleList(
```

So I think it would be safe to just add *zero* channels to the `Z` dim to make the input compatible with the *tiling* size.  
For example for an input of (1, 256, 256) using a careamist model with tiling of (8, 64, 64), we need to add *7* zero channels to the input:
```python
# _input is (1, 256, 256)
fitted_input = np.concatenate([_input, np.zeros((7, 256, 256))], axis=0)
```
If there weren't tiling constraints, one could easily change the padding of the pooling layer for the `Z=1` inputs to `(1, 0, 0)` to make it work.

**Comment by jdeschamps** [comment]
Revisit this soon

**Comment by diyasrivas** [comment]
I redid some of the testing already mentioned here in this issue. I trained a small 3D N2V model:

- for whole image, Z=5, input = (5, 128, 128), no tiling - raised ValueError
- for tiled, Z=5, input = (5, 128, 128),  tile_size=(4,64,64) - works and output (5, 128, 128)
- for whole image, Z=8, input (8, 128, 128), no tiling - works and output (8, 128, 128)

For the error: "Input data dimension Z (size 5) is not a multiple of 4 (2 to the power of the model depth). If you are training, adjust patch_size. If you are predicting, your input data shape is not compatible, use tiling by passing tile_size. If you are already using tiling, adjust tile_size."

So with the current codebase, small/weird Z doesn't fail confusingly, it seems clear. Tiling as a workaround still works, I don't think setting 0 channels for the Z dim is needed? And Valid Z works with whole-image prediction.

**Comment by jdeschamps** [comment]
Thanks @diyasrivas
